File: 26_length_longest_substring_without_repeating.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 26: Find the length of the longest substring without repeating characters.
Uses sliding window.
*/

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    vector<int> last(256, -1);
    int start = 0, best = 0;
    for (int i = 0; i < s.size(); ++i)
    {
        if (last[(unsigned char)s[i]] >= start)
            start = last[(unsigned char)s[i]] + 1;
        last[(unsigned char)s[i]] = i;
        best = max(best, i - start + 1);
    }
    cout << best << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 26_length_longest_substring_without_repeating.py
"""
Task 26: Find length of the longest substring without repeating characters.
Uses sliding window.
"""
import sys

def length_of_longest(s: str) -> int:
    last = {}
    start = 0; best = 0
    for i,ch in enumerate(s):
        if ch in last and last[ch] >= start:
            start = last[ch] + 1
        last[ch] = i
        best = max(best, i-start+1)
    return best

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    print(length_of_longest(s))

# Time Complexity: O(n)
# Space Complexity: O(k)
